src/ratbag_emu/protocol/base.py

```diff
@@ -213,18 +199,87 @@ def _send_packets(self, packets, total):
         for i in range(total):
             s.enter(next_time, 1, self.send_raw,
                     kwargs={'data': self.create_report(packets[i], 0x11)})
-            next_time += 1 / self.hprof.report_rate
+            next_time += 1 / self.hprof.get_report_rate()
         s.run()
 
+
+class BaseDevice():
+    _verbose = True
+
+    endpoints = dict()
+
     '''
-    Gets the active DPI value
+    Represents an hardware device
     '''
-    def get_dpi(self):
-        return self.hprof.dpi[self.hprof.active_dpi]
+
+    protocol = None
+
+    '''
+    Hardware profile
+
+    Represents the active profile in the hardware.
+    '''
+    hprof = Profile()
+
+    active_profile = None
+    profiles = []
+
+    mouse_endpoint = 0
+    keyboard_endpoint = 0
+    media_endpoint = 0
+
+    def __init__(self, rdesc=None, info=None, name='Generic Device',
+                 shortname='generic'):
+        self.info = info
+        self.name = name
+        self.shortname = shortname
+        self.endpoints[0] = SimulatedDevice(rdesc, info, name, shortname)
+
+        for n, endpoint in self.endpoints.items():
+            endpoint.hprof = self.hprof
+
+    '''
+    Pass to the correct endpoint
+    '''
+    def create_report(self, data, type=None):
+        self.endpoints[self.mouse_endpoint].create_report(data, type)
+
+    def simulate_action(self, actions):
+        self.endpoints[self.mouse_endpoint].simulate_action(actions)
+
+    def send_raw_event(self, data):
+        self.endpoints[self.mouse_endpoint].send_raw(data)
+
+    def dispatch(self):
+        for n, endpoint in self.endpoints.items():
+            endpoint.dispatch()
+
+    def destroy(self):
+        for n, endpoint in self.endpoints.items():
+            endpoint.destroy()
+
+    @property
+    def device_nodes(self):
+        nodes = []
+        for n, endpoint in self.endpoints.items():
+            nodes += endpoint.device_nodes
+        return nodes
+
+    @property
+    def verbose(self):
+        return self._verbose
+
+    @verbose.setter
+    def verbose(self, value):
+        self._verbose = value
+        for n, endpoint in self.endpoints.items():
+            endpoint.verbose = value
```

Comment:
the dict seems to be overkill anyway (see @bentiss comment) but you can run through the values directly with:
```
for endpoint in self.endpoints.values():
....
```